PIO2 fails to build on Summit with IBM XL compiler and -O3 in CFLAGS

Steps to reproduce:
```
module load cmake
git clone https://github.com/E3SM-Project/ParallelIO.git
cd ParallelIO
mkdir build
cd build
CC=mpicc CXX=mpicxx FC=mpif90 cmake -DCMAKE_C_FLAGS:STRING="-O3" -DWITH_NETCDF=OFF -DPnetCDF_PATH="/autofs/nccs-svm1_sw/summit/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/xl-16.1.1-3/parallel-netcdf-1.8.1-mib5e6ivbcesgal3hvzvbwxso3dugyun" ..
make
```
The error mesage is:
```
[ 23%] Building C object src/clib/CMakeFiles/pioc.dir/pio_file.c.o
[ 23%] Building C object src/clib/CMakeFiles/pioc.dir/pioc_support.c.o
/sw/summit/xl/16.1.1-3/xlC/16.1.1/bin/.orig/xlc_r: error: 1501-230 Internal compiler error; please contact your Service Representative. For more information visit:
http://www.ibm.com/support/docview.wss?uid=swg21110810
make[2]: *** [src/clib/CMakeFiles/pioc.dir/pioc_support.c.o] Error 41
make[1]: *** [src/clib/CMakeFiles/pioc.dir/all] Error 2
make: *** [all] Error 2
```
This issue can cause E3SM cases to fail during phase build on Summit, as -O3 flag is always used for IBM compiler in non-DEBUG mode:
```
<compiler COMPILER="ibm">
  <CFLAGS>
    <base> -g -qfullpath -qmaxmem=-1 -qphsinfo </base>
    <append DEBUG="FALSE"> -O3  </append>
    ...
  </CFLAGS>
```

--- a/src/clib/CMakeLists.txt
+++ b/src/clib/CMakeLists.txt
@@ -197,3 +197,12 @@ CHECK_TYPE_SIZE("long long" SIZEOF_LONG_LONG)
 if (NOT ${SIZEOF_SIZE_T} EQUAL ${SIZEOF_LONG_LONG})
   message (FATAL_ERROR "size_t and long long must be the same size!")
 endif ()
+
+# Hack to circumvent IBM XL (16.1.1-3) internal compiler error by
+# disabling optimization for pioc_support.c
+cmake_host_system_information (RESULT FQDN_SITENAME QUERY FQDN)
+if (FQDN_SITENAME MATCHES "^.*summit[.]olcf")
+  if (CMAKE_C_COMPILER_NAME STREQUAL "XL")
+    set_source_files_properties(pioc_support.c PROPERTIES COMPILE_FLAGS "-qnoopt -qnosmp")
+  endif()
+endif()

